### PR TITLE
New design for HRM app

### PR DESCRIPTION
**Pull request by moWerk**
Enhance HRM app with progress arc during measuring phase, larger bpm-count, pulsing heart and lastBpm indicator.
 In close collaboration with @MagneFire and thanks to useful comments in the matrix chat.
 
 As requested here:
 https://github.com/AsteroidOS/asteroid/issues/178

### CODE REVIEW

File: src/main.qml [hunk 1]
```diff
@@ -17,27 +17,205 @@
...
+            }
+        }
+    }
+
+    Timer {
+        id: pulseTimer
+        interval: 1000 / (app.bpm / 30)
+        running: app.bpm
+        repeat: true
+        //triggeredOnStart: true
```
Comment (MagneFire) [inline src/main.qml]: Code in comments, please remove.

File: src/main.qml [hunk 2]
```diff
@@ -17,27 +17,205 @@
...
+                                -app.width * 0.008 :
+                                0
+        font.pixelSize: app.bpm > 0 ?
+                            app.bpm / 100 >= 1 ?
+                                app.height*0.29 :
+                                app.height*0.33 :
+                            app.height*0.06
+        text: app.bpm > 0 ?
+                  app.bpm :
+                  qsTrId("id-measuring")
```
Comment (MagneFire) [inline src/main.qml:133]: Add the line `//% "Measuring..."` above here otherwise the language updating tools will break.

File: src/main.qml [hunk 3]
```diff
@@ -17,27 +17,205 @@
...
+        Timer {
+            id: timer
```
Comment (MagneFire) [inline src/main.qml]: It would probably make sense to rename this `timer` to something more descriptive like `blinkingTimer`.

File: src/main.qml [hunk 4]
```diff
@@ -17,27 +17,205 @@
 
 import QtQuick 2.9
 import QtSensors 5.11
+import QtGraphicalEffects 1.15
```
Comment (MagneFire) [inline src/main.qml]: `QtGraphicalEffects` isn't actually used here right?

File: src/main.qml [hunk 5]
```diff
@@ -17,27 +17,205 @@
...
+    property int pulseWidthArc: pulseToggle ?
+                                 0 :
+                                 -6
+    property int pulseWidth: pulseToggle ?
+                                 app.height * 0.20 :
+                                 app.height * 0.26
+
+    onArcEndChanged: canvas.requestPaint()
+    onBpmChanged: {
+        app.arcEnd = Qt.binding(function() { return arcStartOffset + (360 - arcGapHeart) + 28 + pulseWidthArc })
```
Comment (MagneFire) [inline src/main.qml]: This can also be moved to the onCompleted function if I'm not mistaken (Removing the other `app.arcEnd` line).

**Review by MagneFire — COMMENTED** [review]
This is great work!

One final remark is that you should probably add yourself to the credits list here: https://github.com/AsteroidOS/asteroid-hrm/pull/4/files#diff-6b6066b8c2cfa063e0014db476a9967f054fc913bb71de97644e91e91e3660d3R2


I've made some of the comments on a local qml file:
```
/*
 * Copyright (C) 2019 Florent Revest <revestflo@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import QtQuick 2.9
import QtSensors 5.11
import org.asteroid.controls 1.0
import Nemo.KeepAlive 1.1

Application {
    id: app

    centerColor: "#b04d1c"
    outerColor: "#421c0a"

    readonly property real arcStartOffset: -181
    readonly property real arcGapHeart: 52

    property real arcBpmGap: app.lastBpm > 0 ? 28 : 0
    property real arcBpmOffset: app.lastBpm > app.bpm ? 75 : 15

    property real pulseAnimationDuration: app.bpm === 0 ?
                         100 :
                         1000 / (app.bpm / 20)
    property real arcAnimationDuration: app.bpm === 0 ?
                         30000 :
                         1000 / (app.bpm / 20)


    property real arcEnd: arcStart
    property real arcStart: arcStartOffset + arcGapHeart/2 - pulseWidthArc
    property int bpm: 0
    property int lastBpm: 0
    property real lastBpmAngle: app.lastBpm > app.bpm ? 30 : -30
    property bool pulseToggle: true

    property int pulseWidthArc: pulseToggle ?
                                 0 :
                                 -6
    property int pulseWidth: pulseToggle ?
                                 app.height * 0.20 :
                                 app.height * 0.26

    onArcEndChanged: canvas.requestPaint()

    HrmSensor {
        active: true
        onReadingChanged: {
            app.lastBpm = app.bpm
            app.bpm = reading.bpm
            if (app.lastBpm > 0) {
                DisplayBlanking.preventBlanking = false
            }
        }
    }

    Timer {
        id: pulseTimer
        interval: 1000 / (app.bpm / 30)
        running: app.bpm
        repeat: true
        onTriggered: pulseToggle ?
                         pulseToggle = false :
                         pulseToggle = true
    }

    Behavior on arcBpmGap { NumberAnimation { duration: 100; easing.type: Easing.Linear} }
    Behavior on arcBpmOffset { NumberAnimation { duration: 1000; easing.type: Easing.OutCirc } }
    Behavior on arcEnd { NumberAnimation { duration: arcAnimationDuration; easing.type: app.lastBpm === 0 ? Easing.Linear : Easing.OutInSine } }
    Behavior on arcStart { NumberAnimation { duration: arcAnimationDuration; easing.type: app.lastBpm === 0 ? Easing.Linear : Easing.OutInSine } }
    Behavior on pulseWidth { NumberAnimation { duration: pulseAnimationDuration; easing.type: Easing.OutInSine } }

    Canvas {
        id: canvas
        anchors.fill: parent
        rotation: -90
        opacity: 0.4
        onPaint: {
            var ctx = getContext("2d")
            var x = app.width / 2
            var y = app.height / 2
            var start = Math.PI * (arcStart / 180)
            var end = Math.PI * (arcEnd / 180)
            var gap1 = Math.PI * ((arcStartOffset + arcBpmOffset + 45 - arcBpmGap/2) / 180)
            var gap2 = Math.PI * ((arcStartOffset + arcBpmOffset + 45 + arcBpmGap/2) / 180)

            if (gap1 > end) gap1 = end
            if (gap2 > end) gap2 = end

            ctx.reset()
            ctx.beginPath()
            ctx.lineCap = "round"
            ctx.lineWidth = app.height * 0.034
            ctx.strokeStyle = "#38FF12"
            ctx.arc(x, y, (app.width / 2.9), start, gap1, false)
            ctx.stroke()
            ctx.beginPath()
            ctx.arc(x, y, (app.width / 2.9), gap2, end, false)
            ctx.stroke()
        }
    }

    Label {
        id: bpmText
        anchors.centerIn: parent
        horizontalAlignment: Text.AlignHCenter
        font.letterSpacing: app.bpm > 0 ?
                                -app.width * 0.008 :
                                0
        font.pixelSize: app.bpm > 0 ?
                            app.bpm / 100 >= 1 ?
                                app.height*0.29 :
                                app.height*0.33 :
                            app.height*0.06
        text: app.bpm > 0 ?
                  app.bpm :
                  //% "Measuring..."
                  qsTrId("id-measuring")
        Timer {
            id: blinkingTimer
            interval: 800
            running: true
            repeat: app.bpm === 0
            onTriggered: app.bpm === 0 ?
                             bpmText.opacity = bpmText.opacity === 0 ?
                                 1 :
                                 0 :
                             bpmText.opacity = 1
        }
        Behavior on opacity { NumberAnimation { duration: 600; easing.type: Easing.OutQuad } }

    }

    Text {
        id: lastBpmText
        z: 2
        color: "#ffffff"
        anchors {
            horizontalCenter: app.horizontalCenter
            horizontalCenterOffset: -app.width * 0.345
            verticalCenter: app.verticalCenter
        }
        font.letterSpacing: -heartPicture.width * 0.0018
        font.pixelSize: app.height*0.086
        font.styleName: "Bold"
        text: app.lastBpm > 0 ?
                  app.lastBpm :
                  ""
        Text {
            id: arrowShape
            z: 1
            anchors{
                left: lastBpmText.right
                leftMargin: app.width * 0.02
                verticalCenter: lastBpmText.verticalCenter
            }
            font.family: "Source Sans Pro"
            font.pixelSize: app.height * 0.07
            opacity: 1
            color: "#6638FF12"
            text: lastBpm > 0 ? "\u25B6" : ""
        }
        transform: Rotation {
            origin.x: (app.width * 0.345) + (lastBpmText.width * 0.5)
            origin.y: lastBpmText.height * 0.5
            angle: lastBpmAngle
            Behavior on angle {
                NumberAnimation {
                    duration: 1000
                    easing.type: Easing.OutCirc
                }
            }
        }
    }

    Text {
        id: heartPicture
        anchors{
            horizontalCenter: app.horizontalCenter
            verticalCenter: app.verticalCenter
            verticalCenterOffset: app.height * 0.31
        }
        font.pixelSize: pulseWidth
        opacity: 1
        color: "red"
        text: "\u2764"
        Text {
            anchors {
                centerIn: heartPicture
                verticalCenterOffset: heartPicture.height * 0.018
            }
            font.pixelSize: heartPicture.height * 0.22
            font.letterSpacing: -heartPicture.width * 0.004
            font.styleName: "Bold"
            color: "#ffffffff"
            text: "bpm"
        }
    }
    Component.onCompleted: {
        DisplayBlanking.preventBlanking = true

        app.arcEnd = Qt.binding(function() { return arcStartOffset + (360 - arcGapHeart) + 28 + pulseWidthArc })
    }
}

```

File: src/main.qml [hunk 6]
```diff
@@ -17,27 +17,205 @@
...
+            z: 1
+            anchors{
+                left: lastBpmText.right
+                leftMargin: app.width * 0.02
+                verticalCenter: lastBpmText.verticalCenter
+            }
+            font.family: "Source Sans Pro"
+            font.pixelSize: app.height * 0.07
+            opacity: 1
+            color: "#6638FF12"
```
Comment (moWerk) [inline src/main.qml:173]: `opacity: 1` is redundand here due to color definition with alpha channel.

File: src/main.qml [hunk 7]
```diff
@@ -17,27 +17,205 @@
...
+    Text {
+        id: heartPicture
+        anchors{
+            horizontalCenter: app.horizontalCenter
+            verticalCenter: app.verticalCenter
+            verticalCenterOffset: app.height * 0.31
+        }
+        font.pixelSize: pulseWidth
+        opacity: 1
+        color: "red"
```
Comment (moWerk) [inline src/main.qml]: No need to define \u2764 as red since it already is. Also `opacity: 1` is redundant.

**Comment by MagneFire** [comment]
Thanks for fixing my comments and your own 😜

I don't think I should be the one merging it though (as collaborator on the code). Instead I think @jrtberlin or @FlorentRevest should do this one.

**Comment by MagneFire** [comment]
Love the new color! But don't forget to also change it for the desktop file: https://github.com/AsteroidOS/asteroid-hrm/blob/master/asteroid-hrm.desktop.template#L6

This is normally used in the app launcher. But might not work very well if you use anything other than the default launcher style.

**Comment by moWerk** [comment]
> Love the new color! But don't forget to also change it for the desktop file: https://github.com/AsteroidOS/asteroid-hrm/blob/master/asteroid-hrm.desktop.template#L6
> 
> This is normally used in the app launcher. But might not work very well if you use anything other than the default launcher style.

Now i see. I did not quite undestand when you hinted in matrix already. Thanks.

**Comment by jrtberlin** [comment]
Huge improvement! I don't see anything wrong with it. lgtm

**Comment by FlorentRevest** [comment]
Sounds amazing, I'll need to try it out :D Thanks a lot you two !

**Pull request merged**